### DISCUSSION TITLE
Implementing version control

### DISCUSSION THREAD

**Discussion by waelhu** [sensedeep/dynamodb-onetable#198] [Q&A] [answered]
Recently came across this blog post https://aws.amazon.com/blogs/database/implementing-version-control-using-amazon-dynamodb/ and was wondering what are your thoughts and what approach would you recommend for implementing.

Thanks in advance

**Comment by mobsense** [comment, marked as answer]
Interesting.

Here are a few thoughts:

There are two models:  

* meta
* timestamp

```
models: {
    Meta: {
        pk: {type: String, value: 'Equipment#${id}' },
        sk: {type: String, value: 'Metadata' },
        name: {type: String},
        id: {type: Number },
       ....
   },
    Timestamp: {
        pk: {type: String, value: 'Equipment#${id}' },
        sk: {type: String, value: 'Timestamp#${timestamp}' },
        state: {type: String},     
        id: {type: Number },
        timestamp: { type: String, uuid: 'ulid' },
    }
}
```

Queries:

Fetch entire PK

    results = await table.fetch({pk: `Equipment#{id}`)
    results = await table.groupByType(results)

Fetch meta 

    await Meta.get({id})

Fetch timestamps
    await Timestamp.find({id, timestamp})

For a number based version, the transactional model would be a good fit. Where you would create an additional model for the version counter and do the updates in a transaction as per the doc.

**Reply by waelhu** [reply]
Thanks for your feedback and response. Should have mentioned in initial post that we are looking at implementing the number-based versioning. Thanks for covering both scenarios. I hadn't heard about it until one of Rick Houlihan's videos so I wanted to see how onetable was handling it but I guess it's not really a mainstream requirement everyone would need. I'll start looking at implementation as per your recommendation - I'll take a look at some of your transaction samples and see if I can implement accordingly. 
Thanks again, this is GREAT support.
